server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -543,7 +547,14 @@ private static Map<String, Mapper> buildMergedMappers(
                 Mapper mergeIntoMapper = mergedMappers.get(mergeWithMapper.simpleName());
 
                 if (mergeIntoMapper == null) {
-                    mergedMappers.put(mergeWithMapper.simpleName(), mergeWithMapper);
+                    if (mergeWithMapper instanceof ObjectMapper om) {
+                        addNewObjectMapper(reason, objectMergeContext, mergedMappers, om);
```

Comment:
I think I'd slightly rewrite this to do the common `addFieldIfPossible` where the put is currently done , and only if that one fails, do the conditional fallback to the object mapper specific logic.